### PR TITLE
ivy: add x[i;j;k] indexing syntax, including vector-indexed assignment

### PR DESCRIPTION
**Pull request by rsc**
This PR has two commits. 

The first changes multidimensional indexing syntax from x[i][j][k] to x[i; j; k].
Evaluating the old syntax works the same as before;
the new syntax is more efficient, because it doesn't
make temporary copies of x[i] and x[i][j].
Assigning to the old syntax is now an error,
but the error message shows the new syntax.

The second rewrites indexing to fully support
vectors and matrices as indexes, including in assignments.

See the individual commit messages for details.

### CODE REVIEW

**Review by robpike — CHANGES_REQUESTED** [review]
What happens if I do

a[1 2; 3 4] ?

Also doc.go needs fixing.

But this looks good overall; haven't done a full review yet.

**Comment by robpike** [comment]
P.S. After updating doc.go, run
go generate ./...
to generate the help files

**Comment by rsc** [comment]
Will take a look at doc.go.

```
% git change 86
git-codereview: changed to PR 86.
	ivy: implement multidimensional vector-indexed assignment
% go build
% ./ivy
a = 4 4 rho iota 16

a
 1  2  3  4
 5  6  7  8
 9 10 11 12
13 14 15 16

a[1 2; 3 4]
3 4
7 8

a[1 2][3 4]
index a[1 2][(3)] out of range for shape (2 4)

a[1 2; 3 4 5]
index a[_; (5)] out of range for shape (4 4)

```

The ( ) were not intentional - Int.String must be adding them? - but I kind of like them.

**Comment by rsc** [comment]
Updated doc.go, added a line to demo.

File: value/eval.go [hunk 1]
```diff
@@ -111,85 +111,89 @@ func (op *binaryOp) EvalBinary(c Context, u, v Value) Value {
...
 // index computes u[v].
-// v is known to be an int, vector, or matrix.
-// u could be anything.
-// The rank of the result depends only on the rank of u and v,
+// v is a list of successive semicolon-separated index expressions.
+// The rank of the result depends only on the rank of u and each value in v,
 // not on the actual data.
-// For example, if u is a vector and v is a vector,
-// then u[v] is always a vector, even if v has only one element.
+// For example, if u is a vector and v holds a single a vector,
```
Comment (robpike) [inline value/eval.go]: typo

**Review by robpike — CHANGES_REQUESTED** [review]

**Review by robpike — APPROVED** [review]

**Comment by robpike** [comment]
I'll fix the one tweak in the other.

**Pull request merged**